modules/src/ics24_host/error.rs

```diff
@@ -1,31 +1,33 @@
-use anomaly::{BoxError, Context};
-use thiserror::Error;
+use displaydoc::Display;
+use std::string::String;
 
-pub type ValidationError = anomaly::Error<ValidationKind>;
+pub type ValidationError = ValidationKind;
```

Comment:
Can we also use `define_error!` to redefine the `ValidationError` here?

Reply:
`impl ValidationKind {
    pub fn contains_separator(id: String) -> Self {
        Self::ContainsSeparator { id }
    }

    pub fn invalid_length(id: String, length: usize, min: usize, max: usize) -> Self {
        Self::InvalidLength {
            id,
            length,
            min,
            max,
        }
    }

    pub fn invalid_character(id: String) -> Self {
        Self::InvalidCharacter { id }
    }

    pub fn empty() -> Self {
        Self::Empty
    }

    pub fn chain_id_invalid_format(id: String) -> Self {
        Self::ChainIdInvalidFormat { id }
    }

    pub fn kind(&self) -> &Self {
        &self
    }

}
`
If you define the error handling structure by `define_error!`, but how do you implement some functions for this structure? It seems that `flex-error` does not give some hints.

Reply:
The name of the detail struct has a `Detail` postfix. So if you have `define_error!{ ValidationError; ... }`, then you have a `enum ValidationErrorDetail` defined, and you can define `impl ValidationErrorDetail { ... }` for example.

However in this case the helper functions seems to be just on constructing the errors, which we can replace entirely with the constructor functions defined by `flex-error`. So for example, instead of calling `ValidationKind::contains_separator(id)`, we can instead just use `error::contains_separator_error(id)`.

Reply:
En, I understand.